controllers/applicationvolumereplication_controller.go

```diff
@@ -195,295 +191,471 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+
+			requeue = true
+		}
 	}
 
-	a.log.Info("Subscription is unpaused", "name", subscription.Name)
+	a.log.Info("Completed processing placement for AVR", "name", a.instance.Name, "requeue", requeue)
+
+	return requeue
+}
+
+func (a *AVRInstance) processPlacement() bool {
+	a.log.Info("Process AVR Placement", "name", a.avrPlacementRule)
 
-	exists, err := a.vrgManifestWorkAlreadyExists(subscription)
+	requeue := true
+
+	done, err := a.runPrerequisitesForUserAction()
 	if err != nil {
-		return nil, requeue
+		a.log.Error(err, "processPlacement")
+
+		return requeue
 	}
 
-	if exists {
-		return nil, !requeue
+	if !done {
+		a.log.Info("Ran prerequisites but we'll wait until all outstanding operations complete")
+
+		return requeue
 	}
-	// This subscription is ready for manifest (VRG) creation
-	placementDecision, err := a.processUnpausedSubscription(subscription)
+
+	// Running prerequisits is complete.  That also mean that ManifestWork is applied as well.
+	// Now try to update user placementRule
+	neededUpdate, err := a.reconciler.updateUserPlacementRule(a.avrPlacementRule, a.userPlacementRule)
 	if err != nil {
-		a.log.Error(err, "Failed to process unpaused subscription", "name", subscription.Name)
+		a.log.Error(err, "processPlacement")
 
-		return nil, requeue
+		return requeue
 	}
 
-	a.log.Info(fmt.Sprintf("placementDecisions %+v - requeue: %t", placementDecision, !requeue))
+	// If we needed to update user placement rule, then we need to wait for it
+	// to have a chance to run. Hence, requeueing
+	if neededUpdate {
+		a.log.Info("PlacementRule was updated, but we'll give it a chance to run", "PlRuleDecisions",
+			a.userPlacementRule.Status.Decisions)
+
+		return requeue
+	}
+
+	requeue = a.createVRGManifest()
+
+	a.log.Info(fmt.Sprintf("AVR (%+v)", a.instance))
+	a.log.Info(fmt.Sprintf("Made a decision %+v. Requeue? %v", a.instance.Status.Decision, requeue))
+
+	a.advanceToNextDRState()
+
+	return requeue
+}
+
+func (a *AVRInstance) runPrerequisitesForUserAction() (bool, error) {
+	switch a.instance.Spec.Action {
+	case rmn.ActionFailover:
+		return a.runFailoverPrerequisites()
+	case rmn.ActionFailback:
+		return a.runFailbackPrerequisites()
+	case rmn.ActionRelocate:
+		return a.runRelocatePrerequisites()
+	}
+
+	// Not a failover, a failback, or a relocation.  Must be an initial deployement.
+	a.resetDRState()
 
-	return &placementDecision, !requeue
+	return true, nil
 }
 
-func (a *AVRInstance) isSubsriptionPausedForDR(labels map[string]string) bool {
-	return labels != nil &&
-		labels[RamenDRLabelName] != "" &&
-		strings.EqualFold(labels[RamenDRLabelName], "protected") &&
-		labels[subv1.LabelSubscriptionPause] != "" &&
-		strings.EqualFold(labels[subv1.LabelSubscriptionPause], "true")
+func (a *AVRInstance) runFailoverPrerequisites() (bool, error) {
+	a.log.Info("Processing prerequisites for a failover", "Last State", a.instance.Status.LastKnownDRState)
+	a.setDRState(rmn.FailingOver)
```

Comment:
Will set to failingover without checking if it has already failed over? This is in memory though and not being updated till the end, so possibly state will switch back to failed over as the reconcile progresses.

Reply:
That's right. Although no decision is made with the DRState, for now, it is for logging purposes only, so it may need a bit more work there.